### PR TITLE
add upload from URL for catalog_item resource and allow update name and descriptions

### PR DESCRIPTION
**Pull request by vbauzysvmware**
Dep: https://github.com/vmware/go-vcloud-director/pull/422

* update parts were deprecated methods were used
* new property `ova_url` which allows uploading vApp template from URL
* update of resource now support name and description (previously only meta data was changeable)

### CODE REVIEW

File: vcd/resource_vcd_catalog_media.go [hunk 1]
```diff
@@ -256,25 +258,27 @@ func genericVcdMediaRead(d *schema.ResourceData, meta interface{}, origin string
 	metadata, err := media.GetMetadata()
 	if err != nil {
 		log.Printf("[DEBUG] Unable to find media item metadata: %s", err)
-		return err
+		return diag.Errorf("%s", err)
```
Comment (Didainius) [inline vcd/resource_vcd_catalog_media.go:261]: Not a disaster but you could use `diag` builting function which avoids unnecessary conversion throughout many places in this pr

```suggestion
		return diag.FromErr(err)
```

File: vcd/resource_vcd_catalog_item.go [hunk 2]
```diff
@@ -50,22 +52,28 @@ func resourceVcdCatalogItem() *schema.Resource {
...
 			"ova_path": &schema.Schema{
-				Type:        schema.TypeString,
-				Required:    true,
-				ForceNew:    true,
-				Description: "absolute or relative path to OVA",
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"ova_path", "ova_url"},
+				Description:  "absolute or relative path to OVA",
```
Comment (Didainius) [inline vcd/resource_vcd_catalog_item.go:59]:
```suggestion
				Description:  "Absolute or relative path to OVA",
```

File: vcd/resource_vcd_catalog_item.go [hunk 3]
```diff
@@ -79,37 +87,66 @@ func resourceVcdCatalogItem() *schema.Resource {
...
-		return fmt.Errorf(errorRetrievingOrg, err)
+		return diag.Errorf(errorRetrievingOrg, err)
 	}
 
 	catalogName := d.Get("catalog").(string)
 	catalog, err := adminOrg.GetCatalogByName(catalogName, false)
 	if err != nil {
 		log.Printf("[DEBUG] Error finding Catalog: %#v", err)
-		return fmt.Errorf("error finding Catalog: %#v", err)
+		return diag.Errorf("error finding Catalog: %#v", err)
```
Comment (Didainius) [inline vcd/resource_vcd_catalog_item.go:104]: A good opportunity to cleanup this error. No need for `%#v` here

```suggestion
		return diag.Errorf("error finding Catalog: %s", err)
```

File: vcd/resource_vcd_catalog_item.go [hunk 4]
```diff
@@ -79,37 +87,66 @@ func resourceVcdCatalogItem() *schema.Resource {
...
+		return diagError
+	}
+
+	item, err := catalog.GetCatalogItemByName(itemName, true)
+	if err != nil {
+		return diag.Errorf("error retrieving catalog item %s: %s", itemName, err)
+	}
+	d.SetId(item.CatalogItem.ID)
+
+	log.Printf("[TRACE] Catalog item created: %#v", itemName)
```
Comment (Didainius) [inline vcd/resource_vcd_catalog_item.go:126]:
```suggestion
	log.Printf("[TRACE] Catalog item created: %s", itemName)
```

File: vcd/resource_vcd_catalog_item.go [hunk 5]
```diff
@@ -79,37 +87,66 @@ func resourceVcdCatalogItem() *schema.Resource {
...
+	return resourceVcdCatalogItemRead(ctx, d, meta)
+}
+
+func uploadFile(d *schema.ResourceData, catalog *govcd.Catalog, itemName string) diag.Diagnostics {
+	uploadPieceSize := d.Get("upload_piece_size").(int)
 	task, err := catalog.UploadOvf(d.Get("ova_path").(string), itemName, d.Get("description").(string), int64(uploadPieceSize)*1024*1024) // Convert from megabytes to bytes
 	if err != nil {
 		log.Printf("[DEBUG] Error uploading new catalog item: %#v", err)
-		return fmt.Errorf("error uploading new catalog item: %#v", err)
+		return diag.Errorf("error uploading new catalog item: %#v", err)
```
Comment (Didainius) [inline vcd/resource_vcd_catalog_item.go:141]:
```suggestion
		return diag.Errorf("error uploading new catalog item: %s", err)
```

**Review by Didainius — CHANGES_REQUESTED** [review]
First pass.

File: vcd/resource_vcd_catalog_media.go (hunk 1 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go:259]: good point, started to use but not everywhere yet, so made changes to the whole project.

File: vcd/resource_vcd_catalog_item.go (hunk 2 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go:59]: changed

File: vcd/resource_vcd_catalog_item.go (hunk 3 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go:104]: changed in all places where could, still left in project

File: vcd/resource_vcd_catalog_item.go (hunk 4 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go:126]: changed

File: vcd/resource_vcd_catalog_item.go (hunk 5 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go:141]: fixed with previous comment

File: .changes/v3.6.0/770-features.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+* `vcd_catalog_item` allows using `ovf_url` to upload vApp template from URL [GH-770]
+* `vcd_catalog_item` update allows changing `name` and `description`  [GH-770]
```
Comment (lvirbalas) [inline .changes/v3.6.0/770-features.md:2]: Please reflect this addition in the PR's name and future merge commit message.

BTW:
```suggestion
* `vcd_catalog_item` update allows changing `name` and `description` [GH-770]
```

File: vcd/resource_vcd_catalog_item.go [hunk 7]
```diff
@@ -132,68 +183,81 @@ func resourceVcdCatalogItemCreate(d *schema.ResourceData, meta interface{}) erro
...
-	return err
+	return nil
 }
 
-func resourceVcdCatalogItemDelete(d *schema.ResourceData, meta interface{}) error {
+func resourceVcdCatalogItemDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return deleteCatalogItem(d, meta.(*VCDClient))
 }
 
 // currently updates only metadata
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog_item.go:232]: New functionality description missing?

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: .changes/v3.6.0/770-features.md (hunk 6 above)
Comment (vbauzysvmware) [reply .changes/v3.6.0/770-features.md:2]: fixed, PR renamed

File: vcd/resource_vcd_catalog_item.go (hunk 7 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go]: fixed

**Review by Didainius — APPROVED** [review]

**Review by lvirbalas — APPROVED** [review]

**Pull request merged**